From b03a77d7780e1869b48fb11a67b6d4b77f2f5a3f Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Sun, 9 Nov 2025 14:35:21 +0400
Subject: [PATCH] improvement: better gib starting positions

In the original game, all gibs start flying from one single point
(i.e. `playerOrigin`), which is sometimes noticeable.
Let's position the gibs closer to where they are supposed to be
anatomically. That is, legs go lower, head goes higher.
---
 code/cgame/cg_effects.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/code/cgame/cg_effects.c b/code/cgame/cg_effects.c
index ba60f99e..ab6bb416 100644
--- a/code/cgame/cg_effects.c
+++ b/code/cgame/cg_effects.c
@@ -563,12 +563,20 @@ Generated a bunch of gibs launching out from the bodies location
 #define	GIB_JUMP		250
 void CG_GibPlayer( vec3_t playerOrigin ) {
 	vec3_t	origin, velocity;
+	// See `playerMins`, `playerMaxs`.
+	// TODO we could try to check the actual `mins` and `maxs`
+	// (do we have them available on the client though?),
+	// to account for crounching or for the "lying on the ground dead" state.
+	float playerHeight = 32 - MINS_Z;
+	float bottom = playerOrigin[2] + MINS_Z;
+	float playerRadius = 15;
 
 	if ( !cg_blood.integer ) {
 		return;
 	}
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.95 * playerHeight;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
@@ -584,54 +592,75 @@ void CG_GibPlayer( vec3_t playerOrigin ) {
 	}
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.65 * playerHeight;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	CG_LaunchGib( origin, velocity, cgs.media.gibAbdomen );
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.78 * playerHeight;
+	origin[1] += 0.8 * playerRadius;
+	origin[0] += 0.3 * playerRadius;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	CG_LaunchGib( origin, velocity, cgs.media.gibArm );
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.80 * playerHeight;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	CG_LaunchGib( origin, velocity, cgs.media.gibChest );
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.66 * playerHeight;
+	origin[1] += 0.8 * playerRadius;
+	origin[0] += 0.2 * playerRadius;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	CG_LaunchGib( origin, velocity, cgs.media.gibFist );
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.05 * playerHeight;
+	origin[1] += -0.5 * playerRadius;
+	origin[0] += -0.5 * playerRadius;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	CG_LaunchGib( origin, velocity, cgs.media.gibFoot );
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.65 * playerHeight;
+	origin[1] += -0.6 * playerRadius;
+	origin[0] += -0.2 * playerRadius;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	CG_LaunchGib( origin, velocity, cgs.media.gibForearm );
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.57 * playerHeight;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	CG_LaunchGib( origin, velocity, cgs.media.gibIntestine );
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.42 * playerHeight;
+	origin[1] += 0.5 * playerRadius;
+	origin[0] += 0.1 * playerRadius;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
 	CG_LaunchGib( origin, velocity, cgs.media.gibLeg );
 
 	VectorCopy( playerOrigin, origin );
+	origin[2] = bottom + 0.44 * playerHeight;
+	origin[1] += -0.5 * playerRadius;
+	origin[0] += -0.2 * playerRadius;
 	velocity[0] = crandom()*GIB_VELOCITY;
 	velocity[1] = crandom()*GIB_VELOCITY;
 	velocity[2] = GIB_JUMP + crandom()*GIB_VELOCITY;
-- 
2.43.0

